### PR TITLE
Generate client instances with --exportClient option

### PR DESCRIPTION
**Pull request by lub0v-parsable**
I came across this library and was amazed at the way the generator is written and designed! 

Our use case for client: we have multiple backend microservices, each of them expected to have a client library that will be used both on the server-side and in the frontend + use it in the integration tests.

The existing client generation implementation works great when there is only one backend service, but as mentioned in multiple issues (#428, #588, #465) the way the current generated client is implemented is not quite convenient in the era of microservices because it uses the global `OpenAPI` constant to set the configuration for a server calls which makes it almost impossible to configure for multiple microservices.

---

This change is an _addition_ to the existing generation, that solves multiple issues by creating an instance of the client:
* Ability to use multiple generated clients with different backend services;
* Ability to easily switch between different HTTP clients at the execution time, without the need to _generate_ your own HTTP client function;
* Ability to "see" all existing backend services by typing `myClient.` in the IDE because they all grouped in one place.

The client instance is generated by the provided `--exportClient` option to the generator (without it, the usual static client is generated). Additionally, we can create a custom client name (with the `--name` option) that is useful for the multiple backend services (default `AppClient`).

The generated client class looks like this:
```
export class AppClient {
    readonly organizations: OrganizationsService;
    readonly teams: TeamsService;
    readonly request: BaseHttpRequest;

    constructor(openApiConfig?: OpenAPIConfig, HttpRequest: new (config: OpenAPIConfig) => BaseHttpRequest = XhrHttpRequest) {
        this.request = new HttpRequest({
            BASE: openApiConfig?.BASE ?? 'http://localhost:4000',
            VERSION: openApiConfig?.VERSION ?? 'x.x.x',
            WITH_CREDENTIALS: openApiConfig?.WITH_CREDENTIALS ?? false,
            TOKEN: openApiConfig?.TOKEN,
            USERNAME: openApiConfig?.USERNAME,
            PASSWORD: openApiConfig?.PASSWORD,
            HEADERS: openApiConfig?.HEADERS,
        });
        this.organizations = new OrganizationsService(this.request);
        this.teams = new TeamsService(this.request);
    }
}
```

The usage is simple:

```
const client = new AppClient({ BASE: 'http://some-server-host.com', TOKEN: '1234' });
const orgRes: OrganizationResponse = await client.organizations.createOrganization({
  name: 'MyTestOrg',
  description: 'My Test Org Description',
});
```

To change the httpClient, provide its class name as the second parameter in the constructor. By default, the client uses the specified httpClient with the `--client` option, but you can change the client at the execution time like shown below:

```
const client = new AppClient({ ... }, MyAwesomeHttpClient);
```

It would be useful to generate all httpClient classes and put them under the core module to support use cases when the same client class is used in the frontend and the backend, but I didn't add this in the current PR since the number of changes is a bit high already.

### CODE REVIEW

**Comment by BDQ** [comment]
@lub0v-parsable this is an excellent improvement, I did some testing locally to try and help it get merged. Overall it works great!

One minor thing I've found, if you are not using `tags` in your openapi definition then the resulting client that is generated is broken:

```typescript
export class PirateSDK {
    readonly : Service;
    readonly request: BaseHttpRequest;

    constructor(openApiConfig?: OpenAPIConfig, HttpRequest: new (config: OpenAPIConfig) => BaseHttpRequest = FetchHttpRequest) {
        this.request = new HttpRequest({
            BASE: openApiConfig?.BASE ?? '',
            VERSION: openApiConfig?.VERSION ?? '2021-05-25',
            WITH_CREDENTIALS: openApiConfig?.WITH_CREDENTIALS ?? false,
            TOKEN: openApiConfig?.TOKEN,
            USERNAME: openApiConfig?.USERNAME,
            PASSWORD: openApiConfig?.PASSWORD,
            HEADERS: openApiConfig?.HEADERS,
        });
        this. = new Service(this.request);
    }
}
```

This "tags to services" hasn't been a feature we've been leveraging, I'm wondering if it would make sense for the client to extend `Service` so tagging isn't a hard requirement, so the client could look like:

```typescript

export class PirateSDK extends Service{
    readonly ships: ShipsService;
    readonly request: BaseHttpRequest;

    constructor(openApiConfig?: OpenAPIConfig, HttpRequest: new (config: OpenAPIConfig) => BaseHttpRequest = FetchHttpRequest) {
      const  request = new HttpRequest({
            BASE: openApiConfig?.BASE ?? '',
            VERSION: openApiConfig?.VERSION ?? '2021-05-25',
            WITH_CREDENTIALS: openApiConfig?.WITH_CREDENTIALS ?? false,
            TOKEN: openApiConfig?.TOKEN,
            USERNAME: openApiConfig?.USERNAME,
            PASSWORD: openApiConfig?.PASSWORD,
            HEADERS: openApiConfig?.HEADERS,
        });
        super(request)
        this.request = request
        this.ships = new ShipsService(this.request);
    }
}
```

^ This example shows both, some tagged services (ships) and some not.

**Comment by lub0v-parsable** [comment]
Hey @BDQ, thank you for testing it! I definitely missed that case and just pushed the commit with your suggested change. Even though I don't really like that the AppClient plays two roles in this case, it indeed looks good from the usage perspective.

What I mean is that to follow the single responsibility principle it would be better to have a client similar to this:
```
const client = new AppClient(..);
client.teams.someTeamsMethod();
// the shared methods without tags go under the 'service' or similar property
client.service.someMethod();
```
but the usage is not quite user-friendly, whereas with your suggested change, the usage is nicer:

```
const client = new AppClient(..);
client.teams.someTeamsMethod();
client.someMethod();
```

**Comment by BDQ** [comment]
@lub0v-parsable I tested with the latest changes it works great! Thank you.

@ferdikoomen I think this is really great addition, and I've tested it extensively with my projects and it works great. How can I help get this merged?

File: src/templates/core/BaseHttpRequest.hbs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{{>header}}
+
+import type { ApiRequestOptions } from './ApiRequestOptions';
+import type { ApiResult } from './ApiResult';
```
[the remark this answers is not in the document]
Comment (lub0v-parsable) [reply src/templates/core/BaseHttpRequest.hbs:4]: Thank you @HEllRZA, I'll resolve conflicts again and update it.

**Comment by HEllRZA** [comment]
@ferdikoomen 
Would be great if you find the time to merge this pr. It is really leveraging your library.

**Comment by lub0v-parsable** [comment]
@ferdikoomen what can I do to help move this PR further to a "merged" state? Is there anything you would like to change?

**Comment by ferdikoomen** [comment]
@lub0v-parsable Let me check the branch this weekend and get this merged soon!

**Comment by guysegal** [comment]
This is a great PR, will be very useful for me
Is it expected to be merged soon?

**Comment by lub0v-parsable** [comment]
@ferdikoomen Hi! Is there any progress with this one? How can I help you?

I've spent about six hours in total fixing conflicts already, would really love this to go further before any huge refactoring in the main branch :)

**Comment by asbjornenge** [comment]
We need this 😬💖

**Comment by moretalk** [comment]
Could not wait for this. Merged on my fork with the latest from Master here: https://github.com/Ciptex/openapi-typescript-codegen.

Also fixed a few typing issues around Credentials.

**Comment by jpoul** [comment]
I ended up following moretalk's approach of forking as well. This PR adds some great value to an already awesome project. Looking forward to this being merged.

Big thanks to every single contributor! Great work people!

**Comment by ferdikoomen** [comment]
@lub0v-parsable @BDQ @guysegal @jpoul @moretalk @asbjornenge @HEllRZA and others. I have took the code and idea's from this PR and did a cleanup. The result is now in the development branch. I will also add / port the e2e test and then I think we can close this PR and release a new version.

**Comment by ferdikoomen** [comment]
Just pushed 0.16.0 to NPM, happy testing and please let me know if there is an issue with it. I'm closing this PR @lub0v-parsable

**Comment by BDQ** [comment]
That's great @ferdikoomen thank you for taking the time to get this landed, I'll review my code that used @lub0v-parsable original branch and let you know if I hit any issues.

**Comment by HEllRZA** [comment]
Hi @ferdikoomen  i tested it and it seems to be working fine! (v0.18.0)

Only small issue i had is that i could not access the client.config property anymore, because it is private.

However, the code indentation seems to have suffered:
* The parameters of a service don't have any indentation.
* Parameter descriptions that are multi-line get indented wrongly.
```typescript
    /*
     * Foo Service Description
     * More service description
     * @param myParam1 Some description
 *
 * It contains the UUIDs of the assigned users.
     * @param myParam1 Some description`
     */
    public readSomething(
myParam1: string,
myParam1: string,
): CancelablePromise<Foo> {
        return this.httpRequest.request({
 ```

expected:

```typescript
    /*
     * Foo Service Description
     * More service description
     * @param myParam1 Some description
     *
     * It contains the UUIDs of the assigned users.
     * @param myParam1 Some description`
     */
    public readSomething(
        myParam1: string,
        myParam1: string,
    ): CancelablePromise < Foo > {
        return this.httpRequest.request({
 ```

**Comment by ferdikoomen** [comment]
@HEllRZA Let me have a quick look, could be a small regression

**Comment by ferdikoomen** [comment]
@HEllRZA I will make the `request` and `config` props public, but the indentation seems fine for me. Are you using any specific config or spec?

**Comment by HEllRZA** [comment]
@ferdikoomen 
Ok, made me think ... i generated the code on windows. So i tried on linux, and the indentation is just fine!

**Pull request closed without merging**